File: 백준/레벨Gold III/1516. 게임 개발/게임 개발.ts
// @ts-ignore
import fs = require('fs');

const lines = fs.readFileSync('/dev/stdin').toString().split('\n');

// 인접리스트, 진입차수 초기화
const N = parseInt(lines[0]);
const graph: number[][] = Array.from({ length: N + 1 }, () => []);
const inDegree: number[] = Array(N + 1).fill(0);
const buildTime: number[] = Array(N + 1).fill(0);

for (let v = 1; v < N + 1; v++) {
  const buildInfo = lines[v].split(' ').map(Number);
  const [selfBuild, ...dependencies] = buildInfo;
  buildTime[v] = selfBuild;

  for (const u of dependencies) {
    if (u === -1) break;
    graph[u].push(v);
    inDegree[v]++;
  }
}

// 위상정렬
const q: number[] = [];
const result: number[] = Array(N + 1).fill(0);

for (let i = 1; i < N + 1; i++) {
  if (inDegree[i] === 0) {
    q.push(i);
    result[i] = buildTime[i];
  }
}

while (q.length > 0) {
  const u = q.pop();

  for (const v of graph[u]) {
    inDegree[v]--;
    result[v] = Math.max(result[v], result[u] + buildTime[v]);
    if (inDegree[v] === 0) q.push(v);
  }
}

console.log(result.slice(1).join('\n'));

File: 백준/레벨Gold III/1516. 게임 개발/게임 개발.py
# 스택
from sys import stdin

lines = stdin.read().splitlines()

# 인접리스트, 진입차수 초기화
N = int(lines[0])
graph = [[] for _ in range(N + 1)]
in_degree = [0] * (N + 1)
build_time = [0] * (N + 1)

for v in range(1, N + 1):
    build_info = map(int, lines[v].split())
    build_time[v] = next(build_info)
    for u in build_info:
        if u == -1:
            break
        graph[u].append(v)
        in_degree[v] += 1
    
# 위상정렬
q = []
result = [0] * (N + 1)

for i in range(1, N + 1):
    if in_degree[i] == 0:
        q.append(i)
        result[i] = build_time[i]

while q:
    u = q.pop()
    
    for v in graph[u]:
        in_degree[v] -= 1
        result[v] = max(result[v], result[u] + build_time[v])
        if in_degree[v] == 0:
            q.append(v)

print(*result[1:], sep="\n")